frontend/src/components/common/Resource/PortForward.tsx

```diff
@@ -1,78 +1,167 @@
 import { InlineIcon } from '@iconify/react';
 import { Box, Button, CircularProgress, Tooltip, Typography } from '@material-ui/core';
+import grey from '@material-ui/core/colors/grey';
 import MuiLink from '@material-ui/core/Link';
 import { Alert } from '@material-ui/lab';
 import React from 'react';
 import { useTranslation } from 'react-i18next';
 import helpers from '../../../helpers';
-import { deletePortForward, listPortForward, startPortForward } from '../../../lib/k8s/apiProxy';
+import {
+  listPortForward,
+  startPortForward,
+  stopOrDeletePortForward,
+} from '../../../lib/k8s/apiProxy';
+import { KubeContainer, KubeObject } from '../../../lib/k8s/cluster';
+import Pod from '../../../lib/k8s/pod';
+import Service from '../../../lib/k8s/service';
 import { getCluster } from '../../../lib/util';
 import ActionButton from '../ActionButton';
 
 interface PortForwardProps {
-  isPod: boolean;
-  containerPort: number;
-  namespace?: string;
-  name?: string;
-  isPodRunning?: boolean;
+  containerPort: number | string;
+  resource?: KubeObject;
 }
 
-const PODS_BACKEND_PORTFORWARD_KEY = 'pods';
-const SERVICE_BACKEND_PORTFORWARD_KEY = 'services';
-const PORTFORWARD_STORAGE_KEY_FOR_PODS = 'portforward-pods';
-const PORTFORWARD_STORAGE_KEY_FOR_SERVICES = 'portforward-services';
-
 export interface PortForwardState {
   id: string;
   namespace: string;
   cluster: string;
   port: string;
+  status: string;
+}
+
+export const PORT_FORWARDS_STORAGE_KEY = 'portforwards';
+export const PORT_FORWARD_STOP_STATUS = 'Stopped';
+export const PORT_FORWARD_RUNNING_STATUS = 'Running';
+
+function getPortNumberFromPortName(containers: KubeContainer[], namedPort: string) {
+  let portNumber = 0;
+  containers.every((container: KubeContainer) => {
+    container.ports?.find((port: any) => {
+      if (port.name === namedPort) {
+        portNumber = port.containerPort;
+        return false;
+      }
+    });
+    return true;
+  });
+  return portNumber;
+}
+
+function getPodsSelectorFilter(service?: Service) {
+  if (!service) {
+    return '';
+  }
+  const selector = service?.jsonData.spec?.selector;
+  if (selector) {
+    return Object.keys(service?.jsonData.spec?.selector)
+      .map(item => `${item}=${selector[item]}`)
+      .join(',');
+  }
+  return '';
 }
 
 export default function PortForward(props: PortForwardProps) {
-  const { isPod, containerPort, namespace, name, isPodRunning } = props;
+  const { containerPort, resource } = props;
+  const isPod = resource?.metadata?.kind !== 'Service';
+  const service = !isPod ? (resource as Service) : undefined;
+  const namespace = resource?.metadata?.namespace;
+  const name = resource?.metadata?.name;
   const [isCurrentPodOrServicePortForwarding, setIsCurrentPodOrServicePortForwarding] =
     React.useState(false);
```

Comment:
Is this needed when we can just check if the `portForward` is set?

Reply:
removed